File: app/src/main/java/dev/ragnarok/fenrir/model/Link.java
package dev.ragnarok.fenrir.model;

import android.os.Parcel;


public class Link extends AbsModel {

    public static final Creator<Link> CREATOR = new Creator<Link>() {
        @Override
        public Link createFromParcel(Parcel in) {
            return new Link(in);
        }

        @Override
        public Link[] newArray(int size) {
            return new Link[size];
        }
    };
    private String url;
    private String title;
    private String caption;
    private String description;
    private Photo photo;
    private String preview_photo;
    private int msgId;
    private int msgPeerId;

    public Link() {

    }

    protected Link(Parcel in) {
        super(in);
        url = in.readString();
        title = in.readString();
        caption = in.readString();
        description = in.readString();
        preview_photo = in.readString();
        photo = in.readParcelable(Photo.class.getClassLoader());
        msgId = in.readInt();
        msgPeerId = in.readInt();
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        super.writeToParcel(dest, flags);
        dest.writeString(url);
        dest.writeString(title);
        dest.writeString(caption);
        dest.writeString(description);
        dest.writeString(preview_photo);
        dest.writeParcelable(photo, flags);
        dest.writeInt(msgId);
        dest.writeInt(msgPeerId);
    }

    public String getUrl() {
        return url;
    }

    public Link setUrl(String url) {
        this.url = url;
        return this;
    }

    public String getTitle() {
        return title;
    }

    public Link setTitle(String title) {
        this.title = title;
        return this;
    }

    public String getCaption() {
        return caption;
    }

    public Link setCaption(String caption) {
        this.caption = caption;
        return this;
    }

    public String getDescription() {
        return description;
    }

    public Link setDescription(String description) {
        this.description = description;
        return this;
    }

    public Photo getPhoto() {
        return photo;
    }

    public Link setPhoto(Photo photo) {
        this.photo = photo;
        return this;
    }

    public String getPreviewPhoto() {
        return preview_photo;
    }

    public Link setPreviewPhoto(String photo) {
        preview_photo = photo;
        return this;
    }

    public int getMsgId() {
        return msgId;
    }

    public Link setMsgId(int msgId) {
        this.msgId = msgId;
        return this;
    }

    public int getMsgPeerId() {
        return msgPeerId;
    }

    public Link setMsgPeerId(int msgPeerId) {
        this.msgPeerId = msgPeerId;
        return this;
    }

    @Override
    public int describeContents() {
        return 0;
    }
}

File: app_fenrir/src/main/kotlin/dev/ragnarok/fenrir/model/Link.kt
package dev.ragnarok.fenrir.model

import android.os.Parcel
import android.os.Parcelable
import dev.ragnarok.fenrir.readTypedObjectCompat
import dev.ragnarok.fenrir.writeTypedObjectCompat

class Link : AbsModel {
    var url: String? = null
        private set
    var title: String? = null
        private set
    var caption: String? = null
        private set
    var description: String? = null
        private set
    var photo: Photo? = null
        private set
    var previewPhoto: String? = null
        private set
    var msgId = 0
        private set
    var msgPeerId = 0L
        private set

    constructor()
    internal constructor(parcel: Parcel) {
        url = parcel.readString()
        title = parcel.readString()
        caption = parcel.readString()
        description = parcel.readString()
        previewPhoto = parcel.readString()
        photo = parcel.readTypedObjectCompat(Photo.CREATOR)
        msgId = parcel.readInt()
        msgPeerId = parcel.readLong()
    }

    override fun writeToParcel(parcel: Parcel, flags: Int) {
        parcel.writeString(url)
        parcel.writeString(title)
        parcel.writeString(caption)
        parcel.writeString(description)
        parcel.writeString(previewPhoto)
        parcel.writeTypedObjectCompat(photo, flags)
        parcel.writeInt(msgId)
        parcel.writeLong(msgPeerId)
    }

    @AbsModelType
    override fun getModelType(): Int {
        return AbsModelType.MODEL_LINK
    }

    fun setUrl(url: String?): Link {
        this.url = url
        return this
    }

    fun setTitle(title: String?): Link {
        this.title = title
        return this
    }

    fun setCaption(caption: String?): Link {
        this.caption = caption
        return this
    }

    fun setDescription(description: String?): Link {
        this.description = description
        return this
    }

    fun setPhoto(photo: Photo?): Link {
        this.photo = photo
        return this
    }

    fun setPreviewPhoto(photo: String?): Link {
        previewPhoto = photo
        return this
    }

    fun setMsgId(msgId: Int): Link {
        this.msgId = msgId
        return this
    }

    fun setMsgPeerId(msgPeerId: Long): Link {
        this.msgPeerId = msgPeerId
        return this
    }

    override fun describeContents(): Int {
        return 0
    }

    companion object CREATOR : Parcelable.Creator<Link> {
        override fun createFromParcel(parcel: Parcel): Link {
            return Link(parcel)
        }

        override fun newArray(size: Int): Array<Link?> {
            return arrayOfNulls(size)
        }
    }
}